proposals/dark_mode_extension_icons.md

```diff
@@ -0,0 +1,351 @@
+theme. The set of dark mode icons can and will be made automatically visible
+once the system setting is in dark mode, thereby enhancing the overall browser
+experience.
+
+#### Use Cases
+
+1. Improved icon visibility in the extension toolbar.
+1. Improved icon visibility on management the management and shortcuts pages.
+1. Dark mode icon declarations made possible through the extension manifest.
+1. setIcon() will allow setting of dark and/or light mode specific icons.
+1. Improved icon visibility on context menu (Chrome and Safari reply on default
+icons, only Firefox can specify icons).
+1. Improved icon visibility on side panel (Chrome replies on default icons,
+Firefox can specify icons).
+
+### Known Consumers
+
+The Chromium bug has a significant amount of developer interest.
+
+## Specification
+
+### Schema
+
+// TODO: Obtain consensus on any of the following options before merging.
+
+Incumbent action.setIcon(), for reference.
+```
+action.setIcon({
+  path,
+  tabId
+  imageData,
+});
+```
+
+---
+
+#### Option (donut option)
+
+Don't implement setIcon() for dark mode, as it's used for dynamic icons anyway.
+
+* Pros: Easy to implement.
+* Cons: Developers would need to find their own ways of detecting dark mode.
+
+---
+
+#### Option (combined option)
+
+manifest.json
+```
+{
+  "icon_variants": {
+    "dark": {
+      "128": "128_dark.png"
+    },
+    "light": {
+      "128": "128_light.png"
+    }
+  },
+  "action": {
+    "default_icon": {
+      "128": "128_action.png"
+    },
+    "icon_variants": {
+      "dark": {
+        "128": "128_action_dark.png"
+      },
+      "light": {
+        "128": "128_action_light.png"
+      }
+    }
+  }
+}
+```
+
+action.setIcon()
+```
+action.setIcon({
+  variants: {
+    "dark": {
+      "128": "128_action_dark.png"
+    },
+    "light": {
+      "128": LightImageData
+    }
+  },
+});
+```
+
+Schema
+
+manifest.json
+```
+// `Mode` is dark or light (if either is used, both are expected).
+const Modes: string[] = ["dark", "light];
+const Mode = Modes[Math.floor(Math.random() * Modes.length())];
+
+// Primitive type declaration.
+type Size = string;
+type Path = string;
+
+// Map icon size to path.
+const IconSizeToPath: map<Size, Path>;
+
+// Map mode to a string path or a dictionary mapping icon size to a string path.
+const IconVariants: map<Mode, IconSizeToPath | Path>;
+
+{
+  ...
+  "icon_variants": IconVariants;
+  "action": {
+    ...
+    "icon_variants": IconVariants
+  },
+};
+```
+
+action.setIcon()
+```
+// ImageData is an interface representing canvas element pixels.
+const IconSizeToImageData: map<Size, Path>;
+const IconVariantsWithImageData:
+  map<Mode, IconSizeToPath | IconSizeToImageData | ImageData | Path>;
+const ImageDataDictionary: map<Size, ImageData>;
+
+action.setIcon(
+  ...
+  icon_variants?: IconVariantsWithImageData,
+);
+```
+
+* Pros: Easy for developers to use.
+* Cons: Complex to define and implement.
+
+
+---
+
+#### Option (flat option)
```

Comment:
This is the option I prefer. If we do this one, unknown keys should be non-fatal, so we can expend this in the future without too much issues with backwards compatibility.

Reply:
This is also my preferred approach.

Should we say that you can specify an icon for the same theme twice and the browser will use either the first or last one it can load? For example, it would be nice to support:

```
"icon_variants": [
  {
    "16": "16.svg"
  },
  {
    "16": "16.png"
  },
  {
    "16": "dark16.svg"
    "color_scheme": "dark",
  },
  {
    "16": "dark16.png"
    "color_scheme": "dark",
  }]
```

With the understanding browsers that support .svg would use that icon, and those that don't would fallback to the png.

An alternative would be allowing the key to take an array of possible icon files.

Reply:
That sounds ok: path as a string, not an array.

Reply:
If we are going down the path of first icon win (which seems to be preferred). This would mean browsers should ignore icon objects/groups (We should probably find a proper name for this) which have keys it does not understand. Correct me if I am wrong.